lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -261,13 +263,50 @@ fn connect_rdp_inner(
         Ok(())
     });
 
+    let tdp_sd_info_request = Box::new(move |req: SharedDirectoryInfoRequest| -> RdpResult<()> {
+        debug!("sending: {:?}", req);
+        // Create C compatible string from req.path
+        match CString::new(req.path.clone()) {
+            Ok(c_string) => {
+                unsafe {
+                    // Convert C compatible string into raw pointer for passing into the CGO struct
+                    let c_string = c_string.into_raw();
+                    let err = tdp_sd_info_request(
+                        go_ref,
+                        &mut CGOSharedDirectoryInfoRequest {
+                            completion_id: req.completion_id,
+                            directory_id: req.directory_id,
+                            path: c_string,
+                        },
+                    );
+                    // Retake pointer to free memory
+                    let _ = CString::from_raw(c_string);
+                    if err != CGOErrCode::ErrCodeSuccess {
+                        return Err(RdpError::TryError(String::from(
+                            "call to tdp_sd_info_request failed",
+                        )));
+                    };
+                }
+                return Ok(());
+            }
+            Err(_) => {
+                // TODO(isaiah): change TryError to TeleportError for a generic error caused by Teleport specific code.
+                return Err(RdpError::TryError(String::from(format!(
+                    "path contained characters that couldn't be converted to a C string: {}",
+                    req.path
+                ))));
+            }
+        }
+    });
+
     // Client for the "rdpdr" channel - smartcard emulation and drive redirection.
     let rdpdr = rdpdr::Client::new(
         params.cert_der,
         params.key_der,
         pin,
         params.allow_directory_sharing,
         tdp_sd_acknowledge,
+        tdp_sd_info_request,
```

Comment:
I'm not sure I like the way this is developing. Are we going to end up passing a handling into `Client::new` for every single type of message?

Reply:
Agreed, the callbacks seem to very similar things anyway, having a single Rust handler would be nice.

Reply:
> Are we going to end up passing a handler into Client::new for every single type of message?

Yes that's the plan. Passing them in as handlers is based on how you used `handle_remote_copy`. However it may not be necessary to do it this way, I may be able to just pass the `client_ref` in once at initialization (I'll try it out).

> Can we have a single handler for directory sharing messages that Rust always calls into, and then Go can track the state and decide what actions to do on the TDP side?

I'm not quite seeing how that would work. The calling of these tdp functions is going to be deeply intertwined with our RDP message handling.